### ISSUE TITLE
Postgres 9.4 fallback option?

### ISSUE THREAD

**Issue by jeffdoolittle** [JasperFx/Marten#65]
I understand the choice of 9.5 since upserts will be built in. However, until 9.5 is released perhaps a fallback option could be considered for those on a previous 9.x release. Something like this works:

BEGIN;
LOCK TABLE {0} IN SHARE ROW EXCLUSIVE MODE;
WITH upsert AS (UPDATE {0} SET data=@data WHERE id=@id RETURNING *) 
INSERT INTO {0} (id, data)
SELECT @id, @data WHERE NOT EXISTS (SELECT * FROM upsert);
COMMIT;

Thoughts?

**Comment by jeremydmiller** [comment]
@jeffdoolittle,

I'm game for that. Wanna try it out and submit it? Have something that either detects the postgres version or just set it in code and switch? As of now, the upsert is the only 9.5 specific thing I know of. If we did the partial document update feature I think that would require 9.5 specific operators, but I'm not sure that's a big deal to me. We *could* drop down to PLV8 to do that in 9.4.

**Comment by jeffdoolittle** [comment]
I can take a pass at this if you want to assign the issue to me

**Comment by jeffdoolittle** [comment]
I'm not thrilled with the way I set up the configuration, let me know if you have other ideas or preferences. Also, I'm unclear on how to match the args with a column name (see the "todo" I put in the commit).

Unfortunately I only have VS2012 handy at the moment so I wasn't able to build and run but hopefully this gets a good running start on implementing the fallback mechanism.

**Comment by jeremydmiller** [comment]
@jeffdoolittle I think that's fine for now. We might come up for a different name or syntax for the configuration flag later, but that's nothing.

**Comment by jeffdoolittle** [comment]
The latest commit is working with all existing tests on postgres 9.4 However, I don't have a simple fall back mechanism for all the tests, so you'll see that in the latest commit, I just commented out parts of the `SchemaBuilder` so it always uses the postgres 9.4 fallback.

I fiddled with the `DevelopmentModeRegistry` and set the `DocumentSchema.UsePostgres9_4Fallback` value to 'true' which allows most tests to pass. However, there are a few tests that pass a null `DocumentSchema` into the `SchemaBuilder` so I'm at a bit of an impasse for how to get this across the finish line.
